qa/tasks/cephfs/test_exports.py

```diff
@@ -627,3 +627,372 @@ def test_ephemeral_pin_shrink_mds(self):
         log.info("{0} migrations have occured due to the cluster resizing".format(count))
         # rebalancing from 3 -> 2 may cause half of rank 0/1 to move and all of rank 2
         self.assertLessEqual((count/len(subtrees_old)), (1.0/3.0/2.0 + 1.0/3.0/2.0 + 1.0/3.0)*1.25) # aka .66 with 25% overbudget
+
+class TestBalMask(CephFSTestCase):
+    MDSS_REQUIRED = 3
+    CLIENTS_REQUIRED = 1
+
+    def setUp(self):
+        CephFSTestCase.setUp(self)
+
+        self.fs.set_max_mds(3)
+        self.status = self.fs.wait_for_daemons()
+
+        if self.fs.get_var("max_mds") < 3:
+            self.skipTest("Require three mdss")
+
+        self.fs.set_balance_automate(True)
+
+        self.mount_a.run_shell_payload("mkdir -p 1/a")
+
+    def test_set_to_only_invalid_value_for_subdir(self):
+        """
+        That passing invalid values leads to a command failure.
+        """
+        MAX_MDS = 256
+        invalid_values = list(map(str, [-2,MAX_MDS, MAX_MDS+1]))
+        for value in invalid_values:
+            try:
+                self.mount_a.setfattr("1", "ceph.dir.bal.mask", value)
+            except CommandFailedError as e:
+                self.assertEqual(e.exitstatus, 1)
+
+    def test_set_to_mix_invalid_value_for_subdir(self):
+        """
+        That combining valid and invalid values results in a command failure.
+        """
+        MAX_MDS = 256
+        valid_values = list(map(str, [0, 1]))
+        invalid_values = list(map(str, [-2, MAX_MDS, MAX_MDS+1]))
+        invalid_values = [f'{pair[1]},{pair[1]}' for pair in zip(invalid_values, valid_values)]
+        for value in invalid_values:
+            try:
+                self.mount_a.setfattr("1", "ceph.dir.bal.mask", value)
+            except CommandFailedError as e:
+                self.assertEqual(e.exitstatus, 1)
+    def test_distribute_directory_to_multiple_ranks(self):
+        """
+        That bal.mask distributes a directory to multiple ranks.
+        """
+        self.mount_a.setfattr("1", "ceph.dir.bal.mask", "1,2")
+        for i in range(10):
+            self.mount_a.run_shell_payload(f"mkdir -p 1/{i}")
+            self.mount_a.create_n_files(f"1/{i}/file", 100, sync=True)
+
+        time.sleep(15)
+
+        hit_ranks = set()
+        for rank in [0, 1, 2]:
+            subtrees = self._get_subtrees(status=self.status, rank=rank)
+            for s in subtrees:
+                path = s['dir']['path']
+                if path.startswith('/1') and s['auth_first'] == rank:
+                    hit_ranks.add(rank)
+            time.sleep(5)
+
+        self.assertEqual([1,2], sorted(list(hit_ranks)))
```

Comment:
Applying ```ceph.dir.bal.mask``` to one rank has the same effect as ```ceph.dir.pin```. In the example above, the ```dir/``` directory is moved to rank ```1``` and the balanacer only operates at rank ```1```, so no additional subtree movement occurs. If there is confusion with ```ceph.dir.pin```, how about limiting the use of ```ceph.dir.bal.mask``` to 2 or more ranks?

Reply:
No, let's allow it if for no other reason than it makes programs modifying the mask easier.

I think just adding a note in the docs about this would be helpful.

But my question still is: should we get rid of the special code handling static export pins and have `ceph.dir.pin` translate to `ceph.dir.bal.mask`?

Reply:
Follow-up question: does the new `balance_automate` setting disable `ceph.dir.bal.mask`? Should it?

Reply:
Your idea is that we should keep only the interface of ```ceph.dir.pin``` and integrate the related functions into `ceph.dir.bal.mask`? It's possible, but there may be other things to consider. When only the value ```1``` is stored, if getattr is called through two interfaces, both normally return ```1```. The values ```1,2``` are then stored through ceph.dir.bal.mask. After that, when reading `ceph.dir.pin`, an invalid value error is returned, and when reading` ceph.dir.bal.mask`, ```1,2``` is returned normally. Users may be confused. (If I misunderstand your idea, please comment again.)

Since `ceph.dir.pin` is a function that is very familiar to Ceph users, it would be better to keep it as is. I think it would be good to mention the similarities and differences with` ceph.dir.pin `in the documentation.

In the current implementation, subtree migration by updating `ceph.dir.bal.mask` does not occur when Balance_automate is false. Users may want to stop the migration in some cases, so that feature should be provided.

Reply:
> Your idea is that we should keep only the interface of `ceph.dir.pin` and integrate the related functions into `ceph.dir.bal.mask`? It's possible, but there may be other things to consider. When only the value `1` is stored, if getattr is called through two interfaces, both normally return `1`. The values `1,2` are then stored through ceph.dir.bal.mask. After that, when reading `ceph.dir.pin`, an invalid value error is returned, and when reading` ceph.dir.bal.mask`, `1,2` is returned normally. Users may be confused. (If I misunderstand your idea, please comment again.)

I think that particular issue boils down to whether we should allow both `ceph.dir.pin` and `ceph.dir.bal.mask` on the same inode. I think you could keep `ceph.dir.pin` metadata in the `inode_t` but actually implement it in the backend as a `ceph.dir.bal.mask`.

Anyway, it's just a thought that could be refactored later perhaps.

> Since `ceph.dir.pin` is a function that is very familiar to Ceph users, it would be better to keep it as is. I think it would be good to mention the similarities and differences with`ceph.dir.pin`in the documentation.

Yes, I don't think we would retire the interface.
 
> In the current implementation, subtree migration by updating `ceph.dir.bal.mask` does not occur when Balance_automate is false. Users may want to stop the migration in some cases, so that feature should be provided.

Please add a documentation note.

Reply:
Added.

https://github.com/ceph/ceph/compare/9b04cd83fb615464888de665b9933d37724d4a28..ac1ab14f09a789361f9a5c72e448ae85f5b55fa8#diff-cc9cf0d49df0c5644ecb9c98d1f7ac1d307b05837bfdabc6761c65691c6f8dc1R424